tests/tape/test_tape_measure.py

```diff
@@ -501,3 +512,360 @@ def func():
+        type"""
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @qnode(dev)
+        def func():
+            qml.Hadamard(wires=0)
+            return density_matrix(0), expval(qml.PauliZ(1))
+
+        with pytest.raises(
+            QuantumFunctionError,
+            match="The state respectively density matrix"
+            " cannot be returned in combination"
+            " with other return types",
+        ):
+            func()
+
+    def test_return_with_other_types_mixed(self):
+        """Test that an exception is raised when a state is returned along with another return
+        type"""
+
+        dev = qml.device("default.mixed", wires=2)
+
+        @qnode(dev)
+        def func():
+            qml.Hadamard(wires=0)
+            return density_matrix(0), expval(qml.PauliZ(1))
+
+        with pytest.raises(
+            QuantumFunctionError,
+            match="The state respectively density matrix"
+            " cannot be returned in combination"
+            " with other return types",
+        ):
+            func()
+
+    def test_no_state_capability(self, monkeypatch):
+        """Test if an error is raised for devices that are not capable of returning the state
+        respectively the density matrix. This is tested by changing the capability of
+        default.qubit"""
+        dev = qml.device("default.qubit", wires=2)
+        capabilities = dev.capabilities().copy()
+        capabilities["returns_state"] = False
+
+        @qnode(dev)
+        def func():
+            return density_matrix(0)
+
+        with monkeypatch.context() as m:
+            m.setattr(DefaultQubit, "capabilities", lambda *args, **kwargs: capabilities)
+            with pytest.raises(
+                QuantumFunctionError,
+                match="The current device is not capable" " of returning the state",
+            ):
+                func()
+
+    def test_state_not_supported(self):
+        """Test if an error is raised for devices inheriting from the base Device class,
+        which do not currently support returning the state"""
+        dev = qml.device("default.gaussian", wires=2)
+
+        @qnode(dev)
+        def func():
+            return density_matrix(0)
+
+        with pytest.raises(QuantumFunctionError, match="Returning the state is not supported"):
+            func()
+
+    @pytest.mark.parametrize("wires", [[0, 2, 3, 1], ["a", -1, "b", 1000]])
+    def test_custom_wire_labels(self, wires):
+        """Test if an error is raised when custom wire labels are used"""
+        dev = qml.device("default.qubit", wires=wires)
+
+        @qnode(dev)
+        def func():
+            qml.Hadamard(wires=wires[0])
+            for i in range(3):
+                qml.CNOT(wires=[wires[i], wires[i + 1]])
+            return density_matrix(0)
+
+        with pytest.raises(QuantumFunctionError, match="custom wire labels"):
+            func()
```

Comment:
Awesome tests @rmoyard 💪

Reply:
@josh146 Thank you for the review, I will go through the different points today!